File: java/pe023.java
//file: pe023.java

import java.util.*;

public class pe023 {
  public static int sum_factors(int number) {
    int sum = 1;
    int i;
    for (i = 2; i <= Math.sqrt(number); ++i) {
      if ((number % i) == 0) {
        int divisor = number / i;
        sum += i;
        if (i != divisor)
          sum += divisor;
      }
    }
    return sum;
  }

  public static void main( String[] args ) {
    final int MAX = 28123;
    List<Integer> abundant = new Vector<Integer>();
    Map<Integer, Integer> amap = new HashMap<Integer, Integer>();

    for (int i = 1; i < MAX; ++i) {
      int fsum = sum_factors(i);
      if (i < fsum) {
        abundant.add(i);
          amap.put(i, fsum);
      }
    }

    int sum = 0;
    for (int i = 1; i < abundant.get(0); ++i)
      sum += i;

    for (int i = abundant.get(0); i < MAX; ++i) {
      boolean addi = true;
      Iterator jt = abundant.iterator();
      while (jt.hasNext()) {
        int element = ((Integer) jt.next()).intValue();
        if ((element < i) && (amap.get(i - element) != null)) {
          addi = false;
          break;
        }
      }
      if (addi)
        sum += i;
    }
    System.out.println(sum);
  }
}

File: perl/pe023.pl
#!/usr/bin/perl -w
use strict;

my $MAX = 28123;

my @abundant;
my %abundant;
foreach (1 .. $MAX) {
  my $sf = sum_factors($_);
  if ($_ < $sf) {
    push @abundant, $_;
    $abundant{$_} = $sf;
  }
}

my $sum=0;
foreach my $i (1 .. $abundant[0] - 1) {
  $sum += $i;
}
foreach my $i ($abundant[0] .. $MAX) {
  my $addi = 1;
  foreach my $j (@abundant) {
    last if ($j >= $i);
    if ($abundant{$i - $j}) {
      $addi = 0;
      last;
    }
  }
  if ($addi) {
    $sum += $i;
  }
}
print $sum,"\n";

sub sum_factors {
  my $number = shift;

  my $sum = 1;
  for (my $i = 2; $i <= sqrt($number); $i++) {
    if (!($number % $i)) {
      my $divisor = $number / $i;
      $sum += $i;
      $sum += $divisor if ($i != $divisor);
    }
  }
  return $sum;
}
